File: guests/ocaml/feed.ml
(* The feed scene, OCaml port — guests/rust/feed.rs, tools/scenes/feed.steps. *)

open Kaya_wire
open Kaya_app

type post =
  | Note of { text : string }
  | Todo of { title : string; done_ : bool }
[@@deriving kaya_gen]

let () =
  let app = Kaya_app.create () in

  build app (fun () ->
     let feed = sum_of post_sum in
     let done_count =
       sum_derive feed (fun entries ->
           let n =
             List.length
               (List.filter
                  (fun (_, p) -> match p with Todo { done_; _ } -> done_ | _ -> false)
                  entries)
           in
           Str (Printf.sprintf "%d done" n))
     in
     let on_promote () =
       let entries = sum_items feed in
       match
         List.find_opt (fun (_, p) -> match p with Note _ -> true | _ -> false) entries
       with
       | Some (key, Note { text }) ->
           sum_update feed key (Todo { title = text; done_ = true })
       | _ -> ()
     in
     let on_toggle keys checked =
       (* The match is the refinement and the generated patch witnesses it. *)
       let post = sum_get feed (List.hd keys) in
       match post with
       | Some (Todo _) -> post_todo_patch ~done_:checked feed (List.hd keys)
       | _ -> ()
     in

     let root =
       row
         [
           button ~text:"promote" ~on_click:on_promote;
           label ~bind:done_count;
           post_each feed
             ~note:(fun () -> Tpl.(label ~bind_field:post_note_text ()))
             ~todo:(fun () ->
               Tpl.(
                 row
                   [
                     checkbox ~bind_field:post_todo_done_ ~on_toggle;
                     label ~bind_field:post_todo_title;
                   ]
                   ()));
         ]
         ()
     in
     mount root;
     sum_insert feed (Str "a") (Note { text = "jot one" });
     sum_insert feed (Str "b") (Todo { title = "buy milk"; done_ = false });
     sum_insert feed (Str "c") (Note { text = "jot two" }));

  exit (run app)

File: guests/haskell/feed.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE DeriveAnyClass #-}

-- The feed scene, Haskell port — guests/rust/feed.rs, tools/scenes/feed.steps.

import GHC.Generics (Generic)

import Data.Text (Text)
import KayaApp

data Note = Note {text :: Text}
  deriving stock (Generic)
  deriving anyclass (KayaRecord)

data Todo = Todo {title :: Text, done :: Bool}
  deriving stock (Generic)
  deriving anyclass (KayaRecord)

data Post = PNote Note | PTodo Todo
  deriving stock (Generic)
  deriving anyclass (KayaSum)




main :: IO ()
main = kayaMain $ \app -> do
  buildTx app $ do
    feed <- sumCollectionOf @Post
    doneCount <-
      sumDerive feed $ \entries ->
        let n = length [() | (_, PTodo (Todo _ True)) <- entries]
         in tshow n <> " done"

    let onPromote = submitTx app $ do
          entries <- sumItems feed
          case [(k, note) | (k, PNote note) <- entries] of
            (key, Note t) : _ -> sumUpdate feed key (PTodo (Todo t True))
            [] -> pure ()
        onToggle (key : _) checked = submitTx app $ do
          -- The case is the refinement, and the generated patch witnesses it:
          -- a stale occurrence lands in the other arm.
          entry <- sumGet feed key
          case entry of
            Just p@(PTodo _) ->
              sumPatch feed key p [set (field @"done" @Todo) checked]
            _ -> pure ()
        onToggle [] _ = error "kaya: onToggle's key path is never empty"

    root <-
      row
        [ buttonOn "promote" onPromote,
          labelBound doneCount,
          eachSum feed
            [ sumArm (PNote (Note "")) $ do
                _ <- label (field @"text" @Note)
                pure (),
              sumArm (PTodo (Todo "" False)) $ do
                _ <-
                  rowOf
                    [ checkbox (field @"done" @Todo) onToggle,
                      label (field @"title" @Todo)
                    ]
                pure ()
            ]
        ]
    mount root
    sumInsert feed "a" (PNote (Note "jot one"))
    sumInsert feed "b" (PTodo (Todo "buy milk" False))
    sumInsert feed "c" (PNote (Note "jot two"))
